User e2e test

zsxking opened the pull request:
- Add user e2e test
- Update many of the e2e test setup
- Simplify user and data setup before test
- Fix cloudrun scripts

Fix #44 
Fix #45

josueetcom commented on webui/cypress/README.md:
```diff
@@ -3,15 +3,13 @@
 
-1. Go to [Identity Platform Providers][] to enable "Email/Password" provider.
-1. Go to [Identity Platform Users][] to create a user for e2e test.
+This is not needed if run the e2e test on Cloud Build,
+in which the Cloud Build service account already have permission configured.
```
```suggestion
This is not needed if running the e2e test on Cloud Build,
when the Cloud Build service account already have permission configured.
```

ssmall commented on webui/cypress/cloudbuild.yaml:
```diff
@@ -0,0 +1,37 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+## The steps in this Cloud Build script build and publish the frontend
```
Copy+paste?

ssmall commented on webui/cypress/cloudbuild.yaml:
```diff
@@ -0,0 +1,37 @@
+  # Generate the API client for the backend
+  - name: openapitools/openapi-generator-cli:v4.2.3
+    id: Generate Client
+    args: ['generate', '-g','typescript-angular','-i','openapi.yaml','-o','webui/api-client']# generate User Service client
```
Comment seems incorrect

ssmall commented on webui/cypress/credentials.js:
```diff
@@ -0,0 +1,26 @@
+ // The test will create them at the start of the test run,
+ // and clean up them at the end.
+
+exports.adminEmail = "frontend-e2e-cwa-admin@gmail.com";
+exports.adminPassword = "crcwae2eadmin";
+
+exports.workerEmail = "frontend-e2e-cwa-worker@gmail.com";
+exports.workerPassword = "crcwae2eworker";
```
Unfortunately I don't think we can include these default credentials if they are used for anything in the live application -- it would not pass security review. Would it be possible to instead generate random credentials at runtime?

zsxking replied:
Changed it to auto generate password.

ssmall commented on .github/codecov.yml:
```diff
@@ -13,20 +13,22 @@
 # limitations under the License.
 
 codecov:
-    strict_yaml_branch: "master"
```
I'm assuming you removed `strict_yaml_branch: "master"` for test purposes, but it should be added back before merging.

zsxking replied:
sure, will change it back. btw, why it need to be strict to master?

ssmall replied:
This tells codecov to only use the configuration that is in `master` -- otherwise, every PR could overwrite its own code coverage settings.

ssmall commented on .github/codecov.yml:
```diff
@@ -13,20 +13,22 @@
-                only_pulls: true
-          changes: off
\ No newline at end of file
+  ignore:
+    - "webui/src/app/testing"
+    - "**/*.mock.ts"
+  coverage:
+    range: "80...100"
+    status:
+      project:
+        default:
+          target: auto
+          threshold: 0%
+          base: auto
+          only_pulls: true
+      patch:
+        default:
+          target: 100%
+          threshold: 10%
+          base: auto
+          only_pulls: true
+      changes: off
```
does this belong in `.github`? It seems specific to `webui`

zsxking replied:
I haven't figure this out. Content wise, I think it's can be considered not belongs to `webui`, we can just expand the ignore list to cover other files just like how `.gitignore` is. But the main problem is I still haven't be able to get codecov to detect it, and there aren't a config about where this file should be.

ssmall commented on webui/cypress/README.md:
```diff
@@ -3,15 +3,13 @@
-1. Go to [Identity Platform Providers][] to enable "Email/Password" provider.
-1. Go to [Identity Platform Users][] to create a user for e2e test.
+This is not needed if running the e2e test on Cloud Build,
+when the Cloud Build service account already have permission configured.
```
nit: should be "has permissions configured"

ssmall commented on webui/cypress/integration/features/items.feature:
```diff
@@ -2,7 +2,7 @@ Feature: Item Management
 
   Scenario: Create Item
-    Given I logged in
+    Given I logged in as admin
```
Nice! It's very cool to see this integrated with the e2e tests.

ssmall commented on webui/cypress/integration/features/user.feature:
```diff
@@ -0,0 +1,11 @@
+Feature: User Management
+
+  Scenario: Manage User
```
Can you also add a scenario to test that a user who is not an admin should not be able to edit permissions?

zsxking replied:
added

ssmall commented on webui/package.json:
```diff
@@ -9,7 +9,13 @@
     "cypress": "cypress open",
+    "precypress": "npm run prepareE2e",
+    "postcypress": "npm run cleanup",
+    "prepareE2e": "ts-node cypress/scripts/prepareData.ts",
```
nit: casing of `prepareE2e` seems unusual

zsxking replied:
fixed.

josueetcom commented on webui/cloudbuild.yaml:
```diff
@@ -18,14 +18,19 @@ steps:
+  waitFor: ['-']
+  args: ['generate', '-g','typescript-angular','-i','backend/user-service/user-api.yaml','-o','webui/user-svc-client']
 # build the frontend Angular app and deploy it to GCS
 - name: node:12.16
```
You can add
```yaml
  waitFor: ['-']
```
here too. This step doesn't depend on the previous step

josueetcom commented on webui/cloudbuild.yaml:
```diff
@@ -18,14 +18,19 @@ steps:
 # Generate the API client for the backend
 - name: openapitools/openapi-generator-cli:v4.2.3
   id: Generate Client
```
Maybe update the comment and id here? E.g. 
```suggestion
# Generate the API client for the backend and user service
- name: openapitools/openapi-generator-cli:v4.2.3
  id: Generate API Client
```

ssmall commented on webui/cypress/credentials.template.js:
```diff
@@ -0,0 +1,26 @@
+exports.adminPassword = "${ADMIN_PASSWORD}";
+
+exports.workerEmail = "frontend-e2e-cwa-worker@gmail.com";
+exports.workerPassword = "${WORKDER_PASSWORD}";
```
nit: `WORKDER`?

zsxking replied:
:facepalm:

ssmall commented on webui/cypress/scripts/pass-gen.sh:
```diff
@@ -0,0 +1,25 @@
+# limitations under the License.
+
+
+readonly BASE_PATH=$(dirname "${BASH_SOURCE[0]}")/..
+readonly ADMIN_PASSWORD=$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1)
+readonly WORKDER_PASSWORD=$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1)
+
+cat ${BASE_PATH}/credentials.template.js | \
+sed "s/\${ADMIN_PASSWORD}/${ADMIN_PASSWORD}/g" | \
+sed "s/\${WORKDER_PASSWORD}/${WORKDER_PASSWORD}/g" > ${BASE_PATH}/credentials.js
```
This looks fine as-is, but out of curiosity couldn't the random credential generation just be done directly in Javascript?

zsxking replied:
because the javascript import that file, so to do the generation here, it would need to write that value back, and reimport it somehow. it gets messy. Thus I do it separate to keep it clean.

ssmall commented on webui/cypress/cloudbuild.yaml:
```diff
@@ -0,0 +1,38 @@
+
+## The steps in this Cloud Build script run the frontend e2e test
+
+steps:
+  # Generate API Service client
+  - name: openapitools/openapi-generator-cli:v4.2.3
+    id: Generate Client
+    args: ['generate', '-g','typescript-angular','-i','openapi.yaml','-o','webui/api-client']
+  # Generate User Service client
+  - name: openapitools/openapi-generator-cli:v4.2.3
+    id: Generate User Service Client
+    waitFor: ['-']
+    args: ['generate', '-g','typescript-angular','-i','backend/user-service/user-api.yaml','-o','webui/user-svc-client']
+  # build the frontend Angular app and deploy it to GCS
+  - name: cypress/base
+    id: Prepare
+    entrypoint: npm
+    args: ['install']
+    dir: 'webui'
+  - name: cypress/base
+    waitFor: ['Prepare', 'Generate Client', 'Generate User Service Client']
+    id: Test
+    entrypoint: npm
+    args: ['run', 'e2e', '--', '--config', 'baseUrl=https://${_DOMAIN}']
+    dir: 'webui'
```
Is there a missing step to instantiate the credentials template?

zsxking replied:
the initiation is in npm's pre hooks, so no need to run a separate step here.

zsxking commented:
The report looks wrong because I fixed the project root. The file was shown as `src/app/...` before. With the fix it shows as `webui/src/app`. But the report is not able to coup with that change. But anyway I reverted all codecov related changes, as those are not related to the core of this PR and should not be blocking it.

ssmall commented on .github/infra/codecov.yml:
```diff
@@ -13,20 +13,20 @@
-          changes: off
\ No newline at end of file
+  strict_yaml_branch: "master"
+      coverage:
+          range: "80...100"
+          status:
+            project:
+              default:
+                  target: auto
+                  threshold: 0%
+                  base: auto
+                  only_pulls: true
+            patch:
+              default:
+                  target: 100%
+                  threshold: 10%
+                  base: auto
+                  only_pulls: true
+            changes: off 
```
Looks like some extra whitespace got added here
